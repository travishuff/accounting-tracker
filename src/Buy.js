import React from 'react'
import axios from 'axios'
import {
  Button,
  Col,
  Container,
  Form,
  FormGroup,
  FormInput,
  Row,
} from 'shards-react'

class Buy extends React.Component {
  state = {
    number: '',
    buyDate: '',
  }

  handleNumberChange = event => {
    if (!event.target.value) {
      this.setState({
        number: '',
      })
    } else {
      this.setState({
        number: Number(event.target.value),
      })
    }
  }
  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value,
    })
  }
  handleSubmit = async event => {
    event.preventDefault()

    try {
      const response = await axios.post(
        'http://localhost:8080/api/bananas',
        this.state
      )
      console.log('submitted!', response)
    } catch (error) {
      console.error(error)
    }
  }
  handleReset = () => {
    this.setState({
      number: '',
      buyDate: '',
    })
  }
  onSubmit = event => {
    this.handleSubmit(event)
      .then(this.handleReset())
      .catch(console.error)
  }

  render() {
    const { number, buyDate } = this.state

    return (
      <Container fluid={true}>
        <Row className="row-title">
          <Col>Buy Bananas</Col>
        </Row>
        <Row>
          <Col>
            <Form onSubmit={this.onSubmit}>
              <FormGroup>
                <label htmlFor="number">
                  Number of bananas to buy:
                  <FormInput
                    size="sm"
                    type="number"
                    name="number"
                    value={number}
                    onChange={this.handleNumberChange}
                    onBlur={this.handleNumberChange}
                  />
                </label>
              </FormGroup>
              <FormGroup>
                <label htmlFor="buyDate">
                  Purchase date:
                  <FormInput
                    size="sm"
                    type="date"
                    name="buyDate"
                    value={buyDate}
                    onChange={this.handleChange}
                    onBlur={this.handleChange}
                  />
                </label>
              </FormGroup>
              <Button size="sm">Submit</Button>
              <Button size="sm" theme="danger" onClick={this.handleReset}>
                Reset
              </Button>
            </Form>
          </Col>
        </Row>
      </Container>
    )
  }
}

export default Buy
