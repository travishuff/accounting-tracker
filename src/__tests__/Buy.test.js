import React from 'react'
import { create } from 'react-test-renderer'

import Buy from '../Buy'

// Snapshot test
test('snapshot', () => {
  const c = create(<Buy />)

  expect(c.toJSON()).toMatchSnapshot()
})

// Unit testing
test('handleNumberChange method sets value', () => {
  const c = create(<Buy />)
  const instance = c.getInstance()

  instance.state.number = ''
  const event = {
    target: {
      value: 1,
    },
  }
  instance.handleNumberChange(event)
  expect(instance.state.number).toBe(1)
})

test('handleDateChange method sets value', () => {
  const c = create(<Buy />)
  const instance = c.getInstance()

  instance.state.buyDate = '2019-05-01'
  const event = {
    target: {
      value: '2019-06-15',
    },
  }
  instance.handleDateChange(event)
  expect(instance.state.buyDate).toBe('2019-06-15')
})

test('handleSuccessfulSubmit method resets part of state', () => {
  const c = create(<Buy />)
  const instance = c.getInstance()

  instance.state.number = 22
  instance.state.buyDate = '2019-04-13'
  instance.state.hideError = false
  instance.state.hideSuccess = true
  instance.handleSuccessfulSubmit()
  expect(instance.state.number).toBe('')
  expect(instance.state.buyDate).toBe('')
  expect(instance.state.hideError).toBe(true)
  expect(instance.state.hideSuccess).toBe(false)
})

test('resetForm method resets state', () => {
  const c = create(<Buy />)
  const instance = c.getInstance()
  const event = { preventDefault: () => {} }

  instance.state.number = 22
  instance.state.buyDate = '2019-04-13'
  instance.state.hideError = false
  instance.state.hideSuccess = false
  instance.state.dateError = true
  instance.state.numberError = true
  instance.state.errorMsg = 'Error message here.'
  instance.resetForm(event)
  expect(instance.state.number).toBe('')
  expect(instance.state.buyDate).toBe('')
  expect(instance.state.hideError).toBe(true)
  expect(instance.state.hideSuccess).toBe(true)
  expect(instance.state.dateError).toBe(false)
  expect(instance.state.numberError).toBe(false)
  expect(instance.state.errorMsg).toBe('')
})
