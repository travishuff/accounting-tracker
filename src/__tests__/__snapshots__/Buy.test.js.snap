// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="ui fluid container main-container"
>
  <h1>
    Buy Bananas
  </h1>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="eight wide field"
    >
      <label>
        Quantity:
      </label>
      <div
        className="ui input"
      >
        <input
          name="number"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="0"
          type="number"
          value=""
        />
      </div>
    </div>
    <div
      className="eight wide field"
    >
      <label>
        Purchase date:
      </label>
      <div
        className="ui input"
      >
        <input
          name="buyDate"
          onBlur={[Function]}
          onChange={[Function]}
          type="date"
          value=""
        />
      </div>
    </div>
    <div
      className="ui hidden positive message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          Success
        </div>
        <p>
          You bought bananas!
        </p>
      </div>
    </div>
    <div
      className="ui hidden negative message"
    >
      <div
        className="content"
      >
        <div
          className="header"
        >
          Error
        </div>
        <p>
          
        </p>
      </div>
    </div>
    <div
      className="fields"
    >
      <div
        className="field"
      >
        <button
          className="ui primary button"
          onClick={[Function]}
        >
          Submit
        </button>
      </div>
      <div
        className="field"
      >
        <button
          className="ui red button"
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </div>
  </form>
</div>
`;
